File: src/templates.py
# templates.py
# Template and brand discovery and management.
# Finds installed templates and brands in ~/.local/share/md-docs/,
# reads their metadata, lists what's available, and handles
# installation/updates from the templates repo.
# Brands define visual identity (fonts, weights, colors).
# Templates define document layout (resume, paper, letter, etc.).
# Resolves a template + brand pair into paths for the compiler.

import shutil
import subprocess
import tomllib
from pathlib import Path

from config import load_config, CACHE_DIR

TEMPLATES_REPO = "https://github.com/hendemic/md-docs-templates.git"


def _read_metadata_toml(path: Path) -> dict:
    """Read a metadata.toml file, returning empty dict if missing."""
    toml_path = path / "metadata.toml"
    if not toml_path.is_file():
        return {}
    try:
        with open(toml_path, "rb") as f:
            return tomllib.load(f)
    except (tomllib.TOMLDecodeError, OSError):
        return {}


def get_template_metadata(template_path: Path) -> dict:
    """
    Read metadata from a template directory.

    Returns dict with: id, name, description, default_brand, ignore.
    Falls back to directory name for display name.
    """
    raw = _read_metadata_toml(template_path)
    return {
        "id": template_path.name,
        "name": raw.get("name", template_path.name.replace("_", " ").title()),
        "description": raw.get("description"),
        "default_brand": raw.get("default_brand"),
        "ignore": raw.get("ignore", []),
    }


def get_brand_metadata(brand_path: Path) -> dict:
    """
    Read metadata from a brand directory.

    Returns dict with: id, name, description.
    Falls back to directory name for display name.
    """
    raw = _read_metadata_toml(brand_path)
    return {
        "id": brand_path.name,
        "name": raw.get("name", brand_path.name.replace("_", " ").title()),
        "description": raw.get("description"),
    }


def _scan_dir(base_dir: Path, metadata_fn) -> list[dict]:
    """Scan a directory for subdirectories and read their metadata."""
    if not base_dir.is_dir():
        return []
    results = []
    for child in base_dir.iterdir():
        if child.is_dir() and not child.name.startswith("."):
            results.append(metadata_fn(child))
    return sorted(results, key=lambda m: m["name"])


def list_templates(config: dict | None = None) -> list[dict]:
    """Return metadata for all installed templates (repo + custom)."""
    config = config or load_config()
    seen = {}
    # Repo templates first, custom templates override on name collision
    for d in [config["templates_dir"], config["custom_templates_dir"]]:
        for t in _scan_dir(Path(d), get_template_metadata):
            seen[t["id"]] = t
    return sorted(seen.values(), key=lambda m: m["name"])


def list_brands(config: dict | None = None) -> list[dict]:
    """Return metadata for all installed brands (repo + custom)."""
    config = config or load_config()
    seen = {}
    for d in [config["brands_dir"], config["custom_brands_dir"]]:
        for b in _scan_dir(Path(d), get_brand_metadata):
            seen[b["id"]] = b
    return sorted(seen.values(), key=lambda m: m["name"])


def resolve_template(name: str, config: dict | None = None) -> Path:
    """
    Resolve a template name to its directory path.

    Checks custom dir first so user templates override repo templates.
    Raises FileNotFoundError if the template doesn't exist in either location.
    """
    config = config or load_config()
    custom_dir = Path(config["custom_templates_dir"]) / name
    if custom_dir.is_dir():
        return custom_dir
    repo_dir = Path(config["templates_dir"]) / name
    if repo_dir.is_dir():
        return repo_dir
    raise FileNotFoundError(f"Template '{name}' not found")


def load_modifiers() -> dict:
    """
    Load modifiers.toml from the app repo root.

    Returns the raw dict of modifier definitions keyed by modifier id.
    """
    # modifiers.toml lives at the app repo root (one level up from src/)
    app_root = Path(__file__).parent.parent
    modifiers_path = app_root / "modifiers.toml"
    if not modifiers_path.is_file():
        return {}
    try:
        with open(modifiers_path, "rb") as f:
            return tomllib.load(f)
    except (tomllib.TOMLDecodeError, OSError):
        return {}


def resolve_modifiers(modifiers: dict, ignore_list: list[str]) -> dict:
    """
    Build the effective modifier map for a template.

    For each modifier, if it's in the template's ignore list, substitute
    its on_ignore behavior. Otherwise use its normal latex output.

    Args:
        modifiers: Full modifier definitions from modifiers.toml.
        ignore_list: List of modifier ids the template wants to ignore.

    Returns:
        Dict mapping modifier id to a dict with:
            marker, latex, type
        where latex reflects the on_ignore replacement if ignored.
    """
    resolved = {}
    ignore_set = set(ignore_list)
    for mod_id, mod_def in modifiers.items():
        marker = mod_def.get("marker", "")
        mod_type = mod_def.get("type", "block")

        if mod_id in ignore_set:
            on_ignore = mod_def.get("on_ignore", "remove")
            if on_ignore == "remove":
                latex = ""
            elif on_ignore == "newline":
                latex = " \\\\\n" if mod_type == "inline" else "\n\n"
            else:  # "keep"
                latex = None  # signal to leave marker as-is
        else:
            latex = mod_def.get("latex", "")

        resolved[mod_id] = {
            "marker": marker,
            "latex": latex,
            "type": mod_type,
        }
    return resolved


def resolve_brand(name: str, config: dict | None = None) -> Path:
    """
    Resolve a brand name to its directory path.

    Checks custom dir first so user brands override repo brands.
    Raises FileNotFoundError if the brand doesn't exist in either location.
    """
    config = config or load_config()
    custom_dir = Path(config["custom_brands_dir"]) / name
    if custom_dir.is_dir():
        return custom_dir
    repo_dir = Path(config["brands_dir"]) / name
    if repo_dir.is_dir():
        return repo_dir
    raise FileNotFoundError(f"Brand '{name}' not found")


# ---------------------------------------------------------------------------
# Template repo management (install / update / remove)
# ---------------------------------------------------------------------------

def _repo_cache_dir() -> Path:
    """Return the cache path for the cloned templates repo."""
    return CACHE_DIR / "md-docs-templates"


def _git(args: list[str], cwd: Path | None = None) -> subprocess.CompletedProcess:
    """Run a git command. Raises RuntimeError on failure."""
    result = subprocess.run(
        ["git"] + args,
        cwd=cwd,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        msg = result.stderr.strip() or result.stdout.strip()
        raise RuntimeError(f"git {' '.join(args)} failed: {msg}")
    return result


def clone_templates_repo() -> Path:
    """
    Clone the templates repo into cache.

    Returns the cache path.
    Raises RuntimeError if the repo is already cloned or git fails.
    """
    cache = _repo_cache_dir()
    if cache.is_dir() and (cache / ".git").is_dir():
        raise RuntimeError(
            "Templates already installed. Use 'md-docs templates update' to update."
        )
    cache.parent.mkdir(parents=True, exist_ok=True)
    _git(["clone", TEMPLATES_REPO, str(cache)])
    return cache


def pull_templates_repo() -> Path:
    """
    Pull latest changes in the cached templates repo.

    Returns the cache path.
    Raises RuntimeError if the repo isn't cloned yet or git fails.
    """
    cache = _repo_cache_dir()
    if not (cache / ".git").is_dir():
        raise RuntimeError(
            "Templates not installed. Run 'md-docs templates install' first."
        )
    _git(["pull"], cwd=cache)
    return cache


def sync_from_cache(config: dict) -> tuple[list[str], list[str]]:
    """
    Copy templates and brands from the cached repo into the data dirs.

    Returns (list of template ids copied, list of brand ids copied).
    Raises RuntimeError if the cache doesn't exist.
    """
    cache = _repo_cache_dir()
    if not cache.is_dir():
        raise RuntimeError("Templates cache not found. Run 'md-docs templates install' first.")

    templates_src = cache / "templates"
    brands_src = cache / "brands"
    templates_dst = Path(config["templates_dir"])
    brands_dst = Path(config["brands_dir"])

    templates_dst.mkdir(parents=True, exist_ok=True)
    brands_dst.mkdir(parents=True, exist_ok=True)

    installed_templates = []
    installed_brands = []

    if templates_src.is_dir():
        for child in sorted(templates_src.iterdir()):
            if child.is_dir() and not child.name.startswith("."):
                dst = templates_dst / child.name
                if dst.exists():
                    shutil.rmtree(dst)
                shutil.copytree(child, dst)
                installed_templates.append(child.name)

    if brands_src.is_dir():
        for child in sorted(brands_src.iterdir()):
            if child.is_dir() and not child.name.startswith("."):
                dst = brands_dst / child.name
                if dst.exists():
                    shutil.rmtree(dst)
                shutil.copytree(child, dst)
                installed_brands.append(child.name)

    return installed_templates, installed_brands


def repo_template_ids() -> set[str]:
    """
    Return the set of template IDs that came from the templates repo.

    Reads from the cached clone. Returns empty set if cache doesn't exist.
    """
    cache = _repo_cache_dir()
    templates_src = cache / "templates"
    if not templates_src.is_dir():
        return set()
    return {
        child.name
        for child in templates_src.iterdir()
        if child.is_dir() and not child.name.startswith(".")
    }


def repo_brand_ids() -> set[str]:
    """
    Return the set of brand IDs that came from the templates repo.

    Reads from the cached clone. Returns empty set if cache doesn't exist.
    """
    cache = _repo_cache_dir()
    brands_src = cache / "brands"
    if not brands_src.is_dir():
        return set()
    return {
        child.name
        for child in brands_src.iterdir()
        if child.is_dir() and not child.name.startswith(".")
    }


def remove_template(name: str, config: dict) -> bool:
    """
    Remove a template from the templates dir.

    Only removes templates that came from the repo (not user-added).
    Returns True if removed, False if not found.
    Raises ValueError if the template is user-added.
    """
    template_dir = Path(config["templates_dir"]) / name
    if not template_dir.is_dir():
        return False
    if name not in repo_template_ids():
        raise ValueError(
            f"Template '{name}' was not installed from the templates repo. "
            "Remove it manually if needed."
        )
    shutil.rmtree(template_dir)
    return True


def remove_brand(name: str, config: dict) -> bool:
    """
    Remove a brand from the brands dir.

    Only removes brands that came from the repo (not user-added).
    Returns True if removed, False if not found.
    Raises ValueError if the brand is user-added.
    """
    brand_dir = Path(config["brands_dir"]) / name
    if not brand_dir.is_dir():
        return False
    if name not in repo_brand_ids():
        raise ValueError(
            f"Brand '{name}' was not installed from the templates repo. "
            "Remove it manually if needed."
        )
    shutil.rmtree(brand_dir)
    return True

File: src/app/templates.rs
//! Template and brand discovery, metadata parsing, and repository management.
//!
//! Scans configured directories for templates and brands, reads their metadata,
//! loads the modifier registry, and handles git-based install/update/remove operations.
//!
//! # Directory structure expected
//! ```text
//! templates_dir/
//!   resume-2-col/
//!     template.typ
//!     metadata.toml
//!   resume-ats/
//!     template.typ
//!     metadata.toml
//!
//! brands_dir/
//!   generic/
//!     brand.typ
//!     metadata.toml
//! ```

use std::path::{Path, PathBuf};

use crate::domain::{
    Brand, BrandMetadata, Config, ModifierRegistry, Template, TemplateMetadata,
};

// ---------------------------------------------------------------------------
// Modifier loading (standalone, not tied to TemplateManager)
// ---------------------------------------------------------------------------

/// Load the modifier registry from the embedded modifiers.toml.
///
/// The modifiers.toml file is embedded at compile time via `include_str!()`.
/// Returns a HashMap of modifier id -> ModifierDef.
///
/// This is a standalone function because it deserializes a compile-time-embedded
/// string and does not depend on any runtime state or filesystem paths.
pub fn load_modifiers() -> anyhow::Result<ModifierRegistry> {
    todo!("Deserialize include_str!('../../modifiers.toml') into ModifierRegistry")
}

// ---------------------------------------------------------------------------
// TemplateManager
// ---------------------------------------------------------------------------

/// Manages template and brand discovery, metadata loading, and repository operations.
pub struct TemplateManager {
    /// Directory containing installed templates.
    templates_dir: PathBuf,

    /// Directory containing installed brands.
    brands_dir: PathBuf,
}

impl TemplateManager {
    /// Create a new TemplateManager from the resolved config.
    pub fn new(_config: &Config) -> Self {
        todo!("Initialize with templates_dir and brands_dir from config")
    }

    // -----------------------------------------------------------------------
    // Discovery
    // -----------------------------------------------------------------------

    /// Discover all installed templates by scanning the templates directory.
    ///
    /// Each subdirectory with a `template.typ` file is considered a template.
    /// Returns templates sorted by name.
    pub fn discover_templates(&self) -> anyhow::Result<Vec<Template>> {
        todo!("Scan templates_dir, read metadata.toml from each subdir, build Template structs")
    }

    /// Discover all installed brands by scanning the brands directory.
    ///
    /// Each subdirectory with a `brand.typ` file is considered a brand.
    /// Returns brands sorted by name.
    pub fn discover_brands(&self) -> anyhow::Result<Vec<Brand>> {
        todo!("Scan brands_dir, read metadata.toml from each subdir, build Brand structs")
    }

    /// Resolve a template by name.
    ///
    /// Looks up the template in the templates directory.
    /// Returns an error if the template is not found.
    pub fn resolve_template(&self, _name: &str) -> anyhow::Result<Template> {
        todo!("Check templates_dir/name exists, read metadata, return Template")
    }

    /// Resolve a brand by name.
    ///
    /// Looks up the brand in the brands directory.
    /// Returns an error if the brand is not found.
    pub fn resolve_brand(&self, _name: &str) -> anyhow::Result<Brand> {
        todo!("Check brands_dir/name exists, read metadata, return Brand")
    }

    // -----------------------------------------------------------------------
    // Metadata loading
    // -----------------------------------------------------------------------

    /// Read a template's metadata.toml file.
    ///
    /// Falls back to sensible defaults if the file is missing or malformed.
    fn read_template_metadata(_template_dir: &Path) -> anyhow::Result<TemplateMetadata> {
        todo!("Read template_dir/metadata.toml, deserialize into TemplateMetadata")
    }

    /// Read a brand's metadata.toml file.
    ///
    /// Falls back to sensible defaults if the file is missing or malformed.
    fn read_brand_metadata(_brand_dir: &Path) -> anyhow::Result<BrandMetadata> {
        todo!("Read brand_dir/metadata.toml, deserialize into BrandMetadata")
    }

    // -----------------------------------------------------------------------
    // Repository management
    // -----------------------------------------------------------------------

    /// Install templates from a git repository.
    ///
    /// Clones the repository to a cache directory, then copies templates
    /// and brands into the configured data directories.
    pub fn install_repo(&self, _repo_url: &str) -> anyhow::Result<()> {
        todo!("Git clone repo_url to cache, copy templates/ and brands/ to data dirs")
    }

    /// Update installed templates by pulling the latest from the cached repository.
    ///
    /// If `name` is Some, only updates the specified template.
    /// Otherwise updates all templates from the cached repo.
    pub fn update_repo(&self, _name: Option<&str>) -> anyhow::Result<()> {
        todo!("Git pull in cached repo, re-copy templates and brands to data dirs")
    }

    /// Remove a template by name.
    ///
    /// Only removes templates that were installed from a repository.
    /// Returns an error if the template was user-added (not from a repo).
    pub fn remove_template(&self, _name: &str) -> anyhow::Result<()> {
        todo!("Verify template is repo-managed, remove its directory")
    }

    /// Return the cache directory path for a cloned template repository.
    fn cache_dir() -> PathBuf {
        todo!("Return XDG_CACHE_HOME/md-docs/repos or similar")
    }
}
